state/virtualhostkeys.go

```diff
@@ -0,0 +1,109 @@
+package state
+
+import (
+	"crypto/ed25519"
+	"crypto/rand"
+	"encoding/pem"
+
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/errors"
+	"github.com/juju/juju/core/virtualhostkeys"
+	"github.com/juju/mgo/v3"
+	"github.com/juju/mgo/v3/txn"
+)
+
+// VirtualHostKey represents the state of a virtual host key.
+type VirtualHostKey struct {
+	st  *State
+	doc virtualHostKeyDoc
+}
+
+type virtualHostKeyDoc struct {
+	DocId   string `bson:"_id"`
+	Id      string `bson:"id"`
+	HostKey []byte `bson:"hostkey"`
+}
+
+// Id returns the virtual host key's ID.
+func (s *VirtualHostKey) Id() string {
+	return s.doc.Id
+}
+
+// HostKey returns the virtual host key.
+func (s *VirtualHostKey) HostKey() []byte {
+	return s.doc.HostKey
+}
+
+func newVirtualHostKeyDoc(hostKeyID string) (virtualHostKeyDoc, error) {
+	_, privateKey, err := ed25519.GenerateKey(rand.Reader)
```

Comment:
Don't generate host keys at state layer, we should generate it outside of our persistence layer **to be** persisted, rather than persistence with generation.

Reply:
I agree, I thought about that, but because these are "virtual" keys, there is no object who has responsibility for the generation of the host key. So it's something of an auto-generated field and in that view, belongs in the persistence layer, like an auto-generated ID or auto-generated date-time.

Reply:
Alex is correct, the state layer is just for persistence. Ultimately there will be a domain service responsible for generating the keys. In the interim, that sort of domain logic tends to accumulate in the relevant facade.